File: web/src/parseSchema.ts
import convertValue from "./convertValue";
import ArraySchema from "./types/ArraySchema";
import DefSchema from "./types/DefSchema";
import FieldSchema from "./types/FieldSchema";
import MixSchema from "./types/MixSchema";
import ObjectSchema from "./types/ObjectSchema";
import ParseException from "./types/ParseException";
import ParseFailure from "./types/ParseFailure";
import ParseStatus from "./types/ParseStatus";
import ParseSuccess from "./types/ParseSuccess";
import PropsSchema from "./types/PropsSchema";
import RefSchema from "./types/RefSchema";
import Schema from "./types/Schema";
import SchemaValue from "./types/SchemaValue";
import UnionSchema from "./types/UnionSchema";
import { accept, expect, trim } from "./utils";
import validators from "./validators";

interface ParseSchemaStatus extends ParseStatus {
	description: string;
	def: number;
	mix: number;
	any: number;
	refs: Set<string>;
}

/**
 * Parses some structured data from a string into an object.
 *
 * @param input The input string.
 */
export default function parseSchema(input: string): ParseSuccess<Schema> | ParseFailure {
	let status: ParseSchemaStatus = {
		input,
		i: 0,
		errors: [],
		description: "",
		def: 1,
		mix: 1,
		any: 1,
		refs: new Set(),
	};

	trim(status);

	try {
		if (accept("{", status)) {
			const data = parseObject(status);
			if (status.errors.length === 0) {
				return {
					ok: true,
					data,
				};
			} else {
				return { ok: false, errors: status.errors };
			}
		} else {
			status.errors.push({
				message: `Expected '{' but found '${status.input[status.i]}'`,
				index: 0,
				length: 1,
			});
			throw new ParseException();
		}
	} catch (ex) {
		// Handle our errors and throw others
		if (ex instanceof ParseException) {
			return {
				ok: false,
				errors: status.errors,
			};
		} else {
			throw ex;
		}
	}
}

function parseObject(status: ParseSchemaStatus): Schema {
	let result: Schema = {};
	const start = status.i;
	while (true) {
		trim(status);
		if (accept("}", status)) {
			break;
		} else if (status.i === status.input.length || accept("]", status)) {
			status.errors.push({
				message: "Schema object not closed",
				index: start,
				length: 1,
			});
			throw new ParseException();
		}

		parseField(result, status);

		trim(status);
		accept(",", status);
	}
	return result;
}

function parseArray(status: ParseSchemaStatus): SchemaValue {
	trim(status);
	const start = status.i;

	if (accept("]", status)) {
		status.errors.push({
			message: "Schema array empty",
			index: start,
			length: status.i - start,
		});
	} else if (accept("}", status)) {
		status.errors.push({
			message: "Schema array not closed",
			index: start,
			length: 1,
		});
		throw new ParseException();
	}

	let value = parseValue(status);

	trim(status);
	if (status.i === status.input.length || !accept("]", status)) {
		status.errors.push({
			message: "Schema array not closed",
			index: start,
			length: 1,
		});
		throw new ParseException();
	}

	return value;
}

function parseField(result: Record<string, SchemaValue>, status: ParseSchemaStatus): void {
	trim(status);

	const start = status.i;

	// Check for comments
	if (accept("#", status)) {
		const addDescription = accept("#", status);
		for (; status.i < status.input.length; status.i++) {
			if (status.input[status.i] === "\n") {
				break;
			}
		}
		if (addDescription) {
			status.description += status.input.substring(start + 2, status.i);
		}
		return;
	}

	// Check for macros
	if (accept("@", status)) {
		// Consume until space or `(`
		const start = status.i;
		while (/[^\s(]/.test(status.input[status.i])) {
			status.i++;
		}
		const macro = status.input.substring(start, status.i);
		trim(status);
		expect("(", status);
		switch (macro) {
			case "def": {
				trim(status);
				const start = status.i;
				while (status.i < status.input.length && /[^)]/.test(status.input[status.i])) {
					status.i++;
				}
				const ref = status.input.substring(start, status.i).trim();
				// TODO: Should allow any valid prop name, including surrounded by quotes
				if (/[:\s]/.test(ref)) {
					status.errors.push({
						message: `Invalid reference name '${ref}'`,
						index: start,
						length: ref.length,
					});
				}
				status.i++;
				trim(status);
				expect(":", status);
				trim(status);
				expect("{", status);
				status.refs.add(ref);
				const defSchema: DefSchema = {
					type: "def",
					name: ref,
					inner: parseObject(status),
				};
				result[`def$${status.def++}`] = defSchema;
				break;
			}
			case "mix": {
				trim(status);
				const mixSchema: MixSchema = {
					type: "mix",
					inner: [],
				};
				while (true) {
					trim(status);
					if (accept("{", status)) {
						mixSchema.inner.push(parseObject(status));
					} else {
						const start = status.i;
						while (status.i < status.input.length && !/[|)]/.test(status.input[status.i])) {
							status.i++;
						}
						const ref = status.input.substring(start, status.i).trim();
						if (status.refs.has(ref)) {
							const refSchema: RefSchema = {
								type: "ref",
								inner: ref,
							};
							const refResult: Record<string, RefSchema> = {};
							refResult["ref$1"] = refSchema;
							mixSchema.inner.push(refResult);
						} else {
							status.errors.push({
								message: `Unknown reference: '${ref}'`,
								index: start,
								length: ref.length,
							});
						}
					}
					trim(status);
					if (!accept("|", status)) {
						break;
					}
				}
				expect(")", status);
				result[`mix$${status.mix++}`] = mixSchema;
				break;
			}
			case "props": {
				trim(status);
				// Consume until space or `)`
				const start = status.i;
				let level = 1;
				for (; status.i < status.input.length; status.i++) {
					const char = status.input[status.i];
					if (char === "(" && status.input[status.i - 1] !== "\\") {
						level++;
					} else if (char === ")" && status.input[status.i - 1] !== "\\") {
						level--;
						if (level === 0) break;
					} else if (/\s/.test(char)) {
						break;
					}
				}
				const pattern = status.input.substring(start, status.i);
				trim(status);
				expect(")", status);
				trim(status);
				expect(":", status);
				const anyResult: PropsSchema = {
					type: pattern,
					inner: parseValue(status),
				};
				result[`props$${status.mix++}`] = anyResult;
				break;
			}
			default: {
				// Consume until `)`
				const start = status.i - macro.length;
				let level = 1;
				for (; status.i < status.input.length; status.i++) {
					const char = status.input[status.i];
					if (char === "(" && status.input[status.i - 1] !== "\\") {
						level++;
					} else if (char === ")" && status.input[status.i - 1] !== "\\") {
						level--;
						if (level === 0) break;
					}
				}
				status.errors.push({
					message: `Unknown macro: '${macro}'`,
					index: start,
					length: macro.length,
				});
			}
		}
		return;
	}

	let name = "";
	if (accept('"', status)) {
		name = parseString(status);
	} else {
		if (/[^a-zA-Z_]/.test(status.input[status.i])) {
			status.errors.push({
				message: "Field must start with quote or alpha",
				index: start,
				length: 1,
			});
			throw new ParseException();
		}
		status.i++;
		while (/[a-zA-Z0-9_]/.test(status.input[status.i])) {
			status.i++;
		}
		name = status.input.substring(start, status.i);
	}

	trim(status);
	expect(":", status);

	result[name] = parseValue(status);
}

function parseValue(status: ParseSchemaStatus): SchemaValue {
	let value = parseSingleValue(status);

	trim(status);
	if (accept("|", status)) {
		let unionValue: UnionSchema = {
			type: "union",
			inner: [value],
		};
		while (true) {
			trim(status);
			unionValue.inner.push(parseSingleValue(status));
			trim(status);

			if (!accept("|", status)) {
				break;
			}
		}
		value = unionValue;
	}

	return value;
}

function parseSingleValue(status: ParseSchemaStatus): SchemaValue {
	trim(status);
	if (accept("{", status)) {
		const result: ObjectSchema = {
			type: "object",
			inner: parseObject(status),
		};
		return result;
	} else if (accept("[", status)) {
		const result: ArraySchema = {
			type: "array",
			inner: parseArray(status),
		};
		parseValidators(result, status);
		return result;
	} else if (accept('"', status)) {
		return { type: parseString(status, true) };
	} else {
		return parseType(status);
	}
}

function parseType(status: ParseSchemaStatus) {
	// Parse and check the type
	const start = status.i;
	while (status.i < status.input.length && /[^\s|,}\]]/.test(status.input[status.i])) {
		status.i++;
	}
	const type = status.input.substring(start, status.i).trim();
	if (
		!["undef", "null", "bool", "int", "num", "string", "date"].includes(type) &&
		!status.refs.has(type)
	) {
		convertValue(type, start, status.errors);
	}

	// Create the field schema
	let result: FieldSchema = { type };
	if (status.description !== "") {
		result.description = status.description.trim();
		status.description = "";
	}

	parseValidators(result, status);

	return result;
}

function parseValidators(field: FieldSchema, status: ParseSchemaStatus) {
	// Add validators
	trim(status);
	while (status.i < status.input.length && /[^|,}\]]/.test(status.input[status.i])) {
		// Parse and check the validator
		const start = status.i;
		while (status.i < status.input.length && /[^\s|,}\](]/.test(status.input[status.i])) {
			status.i++;
		}
		const validator = status.input.substring(start, status.i);
		if (validators[field.type] && !validators[field.type][validator]) {
			status.errors.push({
				message: `Unsupported validator '${validator}'`,
				index: start,
				length: status.i - start,
			});
		}

		let raw = "true";
		let required = true;

		trim(status);
		if (accept("(", status)) {
			trim(status);
			const start = status.i;
			if (accept('"', status)) {
				raw = parseString(status, true);
				required = convertValue(raw, start, status.errors);
			} else if (accept("/", status)) {
				raw = parseRegex(status);
				required = convertValue(raw, start, status.errors);
			} else {
				// Consume until a space or closing bracket
				const start = status.i;
				while (status.i < status.input.length && /[^\s)]/.test(status.input[status.i])) {
					status.i++;
				}
				raw = status.input.substring(start, status.i);
				required = convertValue(raw, start, status.errors);
			}
			trim(status);
			expect(")", status);
			trim(status);
		}
		(field.validators ??= {})[validator] = { raw, required };
	}
}

function parseString(status: ParseSchemaStatus, withQuotes = false) {
	const start = withQuotes ? status.i - 1 : status.i;
	for (; status.i < status.input.length; status.i++) {
		if (status.input[status.i] === "\\") {
			if (status.input[++status.i] !== '"') {
				status.errors.push({
					message: `Invalid escape sequence '\\${status.input[status.i]}'`,
					index: status.i - 1,
					length: 2,
				});
			}
		} else if (status.input[status.i] === '"') {
			break;
		}
	}
	if (status.i === status.input.length) {
		status.errors.push({
			message: "String not closed",
			index: start,
			length: 1,
		});
		throw new ParseException();
	}
	status.i++;
	const end = withQuotes ? status.i : status.i - 1;
	return status.input.substring(start, end);
}

function parseRegex(status: ParseSchemaStatus) {
	const start = status.i - 1;
	while (
		status.i < status.input.length &&
		!(status.input[status.i] === "/" && status.input[status.i - 1] !== "\\")
	) {
		status.i++;
	}
	if (status.i === status.input.length) {
		status.errors.push({
			message: "Pattern not closed",
			index: start,
			length: 1,
		});
		throw new ParseException();
	}
	while (status.i < status.input.length && /[^\s)]/.test(status.input[status.i])) {
		status.i++;
	}
	return status.input.substring(start, status.i);
}

File: dotnet/Sdnx.Core/ParseSchema.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using Sdnx.Core;

namespace Sdnx.Core
{
    public class ParseSchemaStatus : ParseStatus
    {
        public string Description { get; set; } = string.Empty;
        public int Def { get; set; } = 1;
        public int Mix { get; set; } = 1;
        public int Any { get; set; } = 1;
        public HashSet<string> Refs { get; set; }

        public ParseSchemaStatus() : base()
        {
            Refs = new HashSet<string>();
        }

        public ParseSchemaStatus(string input) : base(input)
        {
            Refs = new HashSet<string>();
        }
    }

    public static class ParseSchema
    {
        /// <summary>
        /// Parses some structured data from a string into an object.
        /// </summary>
        /// <param name="input">The input string.</param>
        public static ParseResult<Schema> Parse(string input)
        {
            var status = new ParseSchemaStatus(input);

            Utils.Trim(status);

            try
            {
                if (Utils.Accept('{', status))
                {
                    var data = ParseObject(status);
                    if (status.Errors.Count == 0)
                    {
                        return new ParseResult<Schema>
                        {
                            Ok = true,
                            Data = new Schema(data)
                        };
                    }
                    else
                    {
                        return new ParseResult<Schema>
                        {
                            Ok = false,
                            Errors = status.Errors
                        };
                    }
                }
                else
                {
                    string found = status.I < status.Input.Length ? status.Input[status.I].ToString() : "undefined";
                    status.Errors.Add(new ParseError(
                        $"Expected '{{' but found '{found}'",
                        0,
                        1
                    ));
                    throw new ParseException();
                }
            }
            catch (ParseException)
            {
                // Handle our errors and throw others
                return new ParseResult<Schema>
                {
                    Ok = false,
                    Errors = status.Errors
                };
            }
        }

        private static Dictionary<string, SchemaValue> ParseObject(ParseSchemaStatus status)
        {
            var result = new Dictionary<string, SchemaValue>();
            int start = status.I;

            while (true)
            {
                Utils.Trim(status);

                if (Utils.Accept('}', status))
                {
                    break;
                }
                else if (status.I >= status.Input.Length || Utils.Accept(']', status))
                {
                    status.Errors.Add(new ParseError(
                        "Schema object not closed",
                        start,
                        1
                    ));
                    throw new ParseException();
                }

                ParseField(result, status);

                Utils.Trim(status);
                Utils.Accept(',', status);
            }

            return result;
        }

        private static SchemaValue ParseArray(ParseSchemaStatus status)
        {
            Utils.Trim(status);
            int start = status.I;

            if (Utils.Accept(']', status))
            {
                status.Errors.Add(new ParseError(
                    "Schema array empty",
                    start,
                    status.I - start
                ));
            }
            else if (Utils.Accept('}', status))
            {
                status.Errors.Add(new ParseError(
                    "Schema array not closed",
                    start,
                    1
                ));
                throw new ParseException();
            }

            var value = ParseValue(status);

            Utils.Trim(status);
            if (status.I >= status.Input.Length || !Utils.Accept(']', status))
            {
                status.Errors.Add(new ParseError(
                    "Schema array not closed",
                    start,
                    1
                ));
                throw new ParseException();
            }

            return value;
        }

        private static void ParseField(Dictionary<string, SchemaValue> result, ParseSchemaStatus status)
        {
            Utils.Trim(status);

            int start = status.I;

            // Check for comments
            if (Utils.Accept('#', status))
            {
                bool addDescription = Utils.Accept('#', status);
                for (; status.I < status.Input.Length; status.I++)
                {
                    if (status.Input[status.I] == '\n')
                    {
                        break;
                    }
                }
                if (addDescription)
                {
                    status.Description += status.Input.Substring(start + 2, status.I - (start + 2));
                }
                return;
            }

            // Check for macros
            if (Utils.Accept('@', status))
            {
                // Consume until space or `(`
                int macroStart = status.I;
                while (status.I < status.Input.Length &&
                       !char.IsWhiteSpace(status.Input[status.I]) &&
                       status.Input[status.I] != '(')
                {
                    status.I++;
                }
                string macro = status.Input.Substring(macroStart, status.I - macroStart);
                Utils.Trim(status);
                Utils.Expect('(', status);

                switch (macro)
                {
                    case "def":
                        Utils.Trim(status);
                        int refNameStart = status.I;
                        while (status.I < status.Input.Length && status.Input[status.I] != ')')
                        {
                            status.I++;
                        }
                        string @ref = status.Input.Substring(refNameStart, status.I - refNameStart).Trim();

                        // TODO: Should allow any valid prop name, including surrounded by quotes
                        if (Regex.IsMatch(@ref, @"[:\s]"))
                        {
                            status.Errors.Add(new ParseError(
                                $"Invalid reference name '{@ref}'",
                                refNameStart,
                                @ref.Length
                            ));
                        }
                        status.I++;
                        Utils.Trim(status);
                        Utils.Expect(':', status);
                        Utils.Trim(status);
                        Utils.Expect('{', status);
                        status.Refs.Add(@ref);

                        var defSchema = new DefSchema
                        {
                            Type = "def",
                            Name = @ref,
                            Inner = ParseObject(status)
                        };
                        result[$"def${status.Def++}"] = defSchema;
                        break;

                    case "mix":
                        Utils.Trim(status);
                        var mixInner = new List<Dictionary<string, SchemaValue>>();

                        while (true)
                        {
                            Utils.Trim(status);
                            if (Utils.Accept('{', status))
                            {
                                mixInner.Add(ParseObject(status));
                            }
                            else
                            {
                                int refNameStart2 = status.I;
                                while (status.I < status.Input.Length &&
                                       status.Input[status.I] != '|' &&
                                       status.Input[status.I] != ')')
                                {
                                    status.I++;
                                }
                                string refName2 = status.Input.Substring(refNameStart2, status.I - refNameStart2).Trim();

                                if (status.Refs.Contains(refName2))
                                {
                                    var refSchema = new RefSchema
                                    {
                                        Type = "ref",
                                        Inner = refName2
                                    };
                                    var refResult = new Dictionary<string, SchemaValue>
                                    {
                                        ["ref$1"] = refSchema
                                    };
                                    mixInner.Add(refResult);
                                }
                                else
                                {
                                    status.Errors.Add(new ParseError(
                                        $"Unknown reference: '{refName2}'",
                                        refNameStart2,
                                        refName2.Length
                                    ));
                                }
                            }
                            Utils.Trim(status);
                            if (!Utils.Accept('|', status))
                            {
                                break;
                            }
                        }
                        Utils.Expect(')', status);

                        var mixSchema = new MixSchema
                        {
                            Type = "mix",
                            Inner = mixInner
                        };
                        result[$"mix${status.Mix++}"] = mixSchema;
                        break;

                    case "props":
                        Utils.Trim(status);
                        // Consume until space or `)`
                        int patternStart = status.I;
                        int level = 1;
                        for (; status.I < status.Input.Length; status.I++)
                        {
                            char ch = status.Input[status.I];
                            if (ch == '(' && status.Input[status.I - 1] != '\\')
                            {
                                level++;
                            }
                            else if (ch == ')' && status.Input[status.I - 1] != '\\')
                            {
                                level--;
                                if (level == 0) break;
                            }
                            else if (char.IsWhiteSpace(ch))
                            {
                                break;
                            }
                        }
                        string pattern = status.Input.Substring(patternStart, status.I - patternStart);
                        Utils.Trim(status);
                        Utils.Expect(')', status);
                        Utils.Trim(status);
                        Utils.Expect(':', status);

                        var propsSchema = new PropsSchema
                        {
                            Type = pattern,
                            Inner = ParseValue(status)
                        };
                        result[$"props${status.Mix++}"] = propsSchema;
                        break;

                    default:
                        // Consume until `)`
                        int macroErrorStart = status.I - macro.Length;
                        level = 1;
                        for (; status.I < status.Input.Length; status.I++)
                        {
                            char ch = status.Input[status.I];
                            if (ch == '(' && status.Input[status.I - 1] != '\\')
                            {
                                level++;
                            }
                            else if (ch == ')' && status.Input[status.I - 1] != '\\')
                            {
                                level--;
                                if (level == 0) break;
                            }
                        }
                        status.Errors.Add(new ParseError(
                            $"Unknown macro: '{macro}'",
                            macroErrorStart,
                            macro.Length
                        ));
                        break;
                }
                return;
            }

            string name;
            if (Utils.Accept('"', status))
            {
                name = ParseString(status);
            }
            else
            {
                if (status.I >= status.Input.Length || !char.IsLetter(status.Input[status.I]) && status.Input[status.I] != '_')
                {
                    status.Errors.Add(new ParseError(
                        "Field must start with quote or alpha",
                        start,
                        1
                    ));
                    throw new ParseException();
                }
                status.I++;
                while (status.I < status.Input.Length &&
                       (char.IsLetterOrDigit(status.Input[status.I]) || status.Input[status.I] == '_'))
                {
                    status.I++;
                }
                name = status.Input.Substring(start, status.I - start);
            }

            Utils.Trim(status);
            Utils.Expect(':', status);

            result[name] = ParseValue(status);
        }

        private static SchemaValue ParseValue(ParseSchemaStatus status)
        {
            var value = ParseSingleValue(status);

            Utils.Trim(status);
            if (Utils.Accept('|', status))
            {
                var unionInner = new List<SchemaValue> { value };
                while (true)
                {
                    Utils.Trim(status);
                    unionInner.Add(ParseSingleValue(status));
                    Utils.Trim(status);

                    if (!Utils.Accept('|', status))
                    {
                        break;
                    }
                }
                value = new UnionSchema
                {
                    Type = "union",
                    Inner = unionInner
                };
            }

            return value;
        }

        private static SchemaValue ParseSingleValue(ParseSchemaStatus status)
        {
            Utils.Trim(status);
            if (Utils.Accept('{', status))
            {
                return new ObjectSchema
                {
                    Type = "object",
                    Inner = ParseObject(status)
                };
            }
            else if (Utils.Accept('[', status))
            {
                var result = new ArraySchema
                {
                    Type = "array",
                    Inner = ParseArray(status)
                };
                return ParseValidators(result, status);
            }
            else if (Utils.Accept('"', status))
            {
                string typeValue = ParseString(status);
                return new FieldSchema
                {
                    Type = typeValue
                };
            }
            else
            {
                return ParseType(status);
            }
        }

        private static SchemaValue ParseType(ParseSchemaStatus status)
        {
            // Parse and check the type
            int start = status.I;
            while (status.I < status.Input.Length &&
                   status.Input[status.I] != ' ' &&
                   status.Input[status.I] != '\t' &&
                   status.Input[status.I] != '|' &&
                   status.Input[status.I] != ',' &&
                   status.Input[status.I] != '}' &&
                   status.Input[status.I] != ']')
            {
                status.I++;
            }
            string type = status.Input.Substring(start, status.I - start).Trim();

            var validTypes = new HashSet<string> { "undef", "null", "bool", "int", "num", "string", "date" };
            if (!validTypes.Contains(type) && !status.Refs.Contains(type))
            {
                ConvertValue.Convert(type, start, status.Errors);
            }

            // Create a field schema
            FieldSchema result = new FieldSchema
            {
                Type = type
            };

            if (!string.IsNullOrEmpty(status.Description))
            {
                result = result with { Description = status.Description.Trim() };
                status.Description = string.Empty;
            }

            return ParseValidators(result, status);
        }

        private static SchemaValue ParseValidators(SchemaValue field, ParseSchemaStatus status)
        {
            // Only certain schema types support validators - skip array, object, def, etc.
            if (field is ArraySchema)
            {
                return field;
            }
            if (field is ObjectSchema)
            {
                return field;
            }
            if (field is DefSchema)
            {
                return field;
            }
            if (field is MixSchema)
            {
                return field;
            }
            if (field is PropsSchema)
            {
                return field;
            }
            if (field is RefSchema)
            {
                return field;
            }
            if (field is UnionSchema)
            {
                return field;
            }

            if (field is not FieldSchema baseField)
            {
                return field;
            }

            // Add validators
            Utils.Trim(status);
            while (status.I < status.Input.Length &&
                   status.Input[status.I] != '|' &&
                   status.Input[status.I] != ',' &&
                   status.Input[status.I] != '}' &&
                   status.Input[status.I] != ']')
            {
                // Parse and check the validator
                int start = status.I;
                while (status.I < status.Input.Length &&
                       status.Input[status.I] != ' ' &&
                       status.Input[status.I] != '\t' &&
                       status.Input[status.I] != '|' &&
                       status.Input[status.I] != ',' &&
                       status.Input[status.I] != '}' &&
                       status.Input[status.I] != ']' &&
                       status.Input[status.I] != '(')
                {
                    status.I++;
                }
                string validator = status.Input.Substring(start, status.I - start);

                if (Validators.All.TryGetValue(baseField.Type, out var typeValidators) &&
                    typeValidators != null &&
                    !typeValidators.ContainsKey(validator))
                {
                    status.Errors.Add(new ParseError(
                        $"Unsupported validator '{validator}'",
                        start,
                        status.I - start
                    ));
                }

                string raw = "true";
                object? required = true;

                Utils.Trim(status);
                if (Utils.Accept('(', status))
                {
                    Utils.Trim(status);
                    int valueStart = status.I;

                    if (Utils.Accept('"', status))
                    {
                        raw = ParseString(status);
                        required = ConvertValue.Convert(raw, start, status.Errors);
                    }
                    else if (Utils.Accept('/', status))
                    {
                        raw = ParseRegex(status);
                        required = ConvertValue.Convert(raw, start, status.Errors);
                    }
                    else
                    {
                        // Consume until a space or closing bracket
                        int valueStart2 = status.I;
                        while (status.I < status.Input.Length &&
                               !char.IsWhiteSpace(status.Input[status.I]) &&
                               status.Input[status.I] != ')')
                        {
                            status.I++;
                        }
                        raw = status.Input.Substring(valueStart2, status.I - valueStart2);
                        required = ConvertValue.Convert(raw, start, status.Errors);
                    }
                    Utils.Trim(status);
                    Utils.Expect(')', status);
                    Utils.Trim(status);
                }

                var validators = baseField.Validators ?? new Dictionary<string, Validator>();
                validators[validator] = new Validator
                {
                    Raw = raw,
                    Required = required
                };
                baseField = baseField with { Validators = validators };
            }

            return baseField;
        }

        private static string ParseString(ParseSchemaStatus status)
        {
            int start = status.I - 1;
            for (; status.I < status.Input.Length; status.I++)
            {
                if (status.Input[status.I] == '\\')
                {
                    status.I++;
                    if (status.I < status.Input.Length && status.Input[status.I] != '"')
                    {
                        status.Errors.Add(new ParseError(
                            $"Invalid escape sequence '\\{status.Input[status.I]}'",
                            status.I - 1,
                            2
                        ));
                    }
                }
                else if (status.Input[status.I] == '"')
                {
                    break;
                }
            }

            if (status.I == status.Input.Length)
            {
                status.Errors.Add(new ParseError(
                    "String not closed",
                    start,
                    1
                ));
                throw new ParseException();
            }
            status.I++;
            return status.Input.Substring(start, status.I - start);
        }

        private static string ParseRegex(ParseSchemaStatus status)
        {
            int start = status.I - 1;
            while (status.I < status.Input.Length &&
                   !(status.Input[status.I] == '/' && status.Input[status.I - 1] != '\\'))
            {
                status.I++;
            }

            if (status.I == status.Input.Length)
            {
                status.Errors.Add(new ParseError(
                    "Pattern not closed",
                    start,
                    1
                ));
                throw new ParseException();
            }

            while (status.I < status.Input.Length &&
                   !char.IsWhiteSpace(status.Input[status.I]) &&
                   status.Input[status.I] != ')')
            {
                status.I++;
            }

            return status.Input.Substring(start, status.I - start);
        }
    }
}
